client/rpc/trace/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+	B: BlockT<Hash = H256> + Send + Sync + 'static,
+	C::Api: BlockBuilder<B, Error = BlockChainError>,
+	C::Api: DebugRuntimeApi<B>,
+	C::Api: EthereumRuntimeRPCApi<B>,
+	A: EthApi,
+{
+	// TODO :
+	// 1. Handle requests and add traces to the cache :
+	//    Cache is a BTreeMap : Block height => Vec of TransactionTrace + expiration time
+	//    No filtering is done in the cache, it stores all traces.
+	//    Existing block in cache get the expiration time bumped.
+	//
+	//    Filtering is done on top :
+	//    1. Apply the filter and return a list of indices to keep
+	//    2. Use the indices to build the filtered vec of traces (with correct pointers)
+	//
+	// 2. Remove expired cache :
+	//    Iterate over each block in the BTreeMap, and remove the entry if expired.
+	//    Question : Is the expiration time and delay between checks configurable ?
+	//    Other idea : Spawn a timer future when updating the cache, providing the height
+	//        of the block. When woken up, check only this block expiration time.
+	//        Will create a future for each request, but is more reactive to cleanup.
+	//        Which is better ?
+
+	// How to get Ethereum block hash and transactions :
+	// EthApi::block_by_number -> Rich<Block>
+	// Block.transactions (Full).hash
+	// Debug RPC impl shows how to get the substrate equivalents for mapping.
+
+	pub fn task(
+		client: Arc<C>,
+		backend: Arc<BE>,
+		eth_api: A,
+	) -> (impl Future<Output = ()>, TraceFilterCacheRequester) {
+		const EXPIRATION_DELAY: Duration = Duration::from_secs(600);
+
+		let (tx, mut rx): (TraceFilterCacheRequester, _) =
+			sp_utils::mpsc::tracing_unbounded("trace-filter-cache-requester");
+
+		let fut = async move {
+			let mut expiration_futures = FuturesUnordered::new();
+			let mut cached_blocks = BTreeMap::<u32, CacheBlock>::new();
+
+			tracing::info!("Begining Trace Filter Cache Task ...");
+
+			'service: loop {
+				select! {
+					req = rx.next() => {
+						if let Some((req, response_tx)) = req {
+							let span = tracing::debug_span!("received request", request = ?req);
+							let _guard = span.enter();
+
+							tracing::trace!("Begining handling request");
+
+							let from_block = match req.from_block {
+								None => 1,
+								Some(RequestBlockId::Number(n)) => n,
+								_ => todo!("support latest/earliest/pending"),
+							};
+
+							let to_block = match req.from_block {
+								None => todo!("support latest"),
+								Some(RequestBlockId::Number(n)) => n,
+								_ => todo!("support latest/earliest/pending"),
+							};
+
+							let from_address = req.from_address.unwrap_or_default();
+							let to_address = req.to_address.unwrap_or_default();
+
+							let range = from_block ..= to_block;
+
+							let block_heights: Vec<u32> = range.clone().collect();
+
+							// Fill cache if needed.
+							for block_height in block_heights.iter() {
+								let cached = cached_blocks.contains_key(block_height);
```

Comment:
If we cache tracing data by block number, couldn't this be problematic in a re-org?

Reply:
Indeed ! I already find it weird that the RPC endpoint expect a block number instead of a block hash. That's also the reason I have not implemented "latest" too. We can make the cache having block-hash keys, and when requesting only getting the `to_block` by number, and then following the chain of parents to get the corresponding blocks and their hashes.

Reply:
Yeah I think that's a good idea, we can use either the ethereum block hash or the substrate block hash as a key.